[RNMobile] Refactor simplify media flow redux store changes

jd-alexander opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
The purpose of this PR is to refactor changes added in https://github.com/WordPress/gutenberg/pull/29546 . A separate PR targeting that PR was utilized to reduce the diff changes the reviewer has to wrangle, so that the review can be done in more seamless manner. The main changes that were made were as a result of this [feedback](https://github.com/WordPress/gutenberg/pull/29546#discussion_r597449997).

- Move the actions, selectors, reducers from the `editor` package to the `block-editor` package. 
- Move the tests of the actions, selectors, reducers from the `editor` package to the `block-editor` package. 
- Move the actions, selectors, and reducer from `native` to `web` to resolve e2e test issues of the base PR.
- Remove string literals being used to identify the data store and used store definition instead.

## Testing 

1. Tap (+) (inserter) – this shows the Block Library sheet
2. Tap Image (or Video or Gallery) to add the block's media placeholder to canvas – which triggers the media upload options sheet. 
3. Choose the media upload option.
4. Select an image.
5. The image is then added to the canvas, uploads succeed/fails.

### Image 🟢 

Android | iOS 
--------|-------
 <kbd><img src="https://user-images.githubusercontent.com/1509205/109728900-11892700-7b85-11eb-9420-7c5818926068.gif" width="320"></kbd>    |       <kbd><img src="https://user-images.githubusercontent.com/1509205/109728904-1221bd80-7b85-11eb-9340-fc4ca0bb1daf.gif" width="320"></kbd>

### Video  🟢 

Android | iOS 
--------|-------
  <kbd><img src="https://user-images.githubusercontent.com/1509205/109728905-12ba5400-7b85-11eb-878f-bfef2ac954df.gif" width="320"></kbd>    |     <kbd><img src="https://user-images.githubusercontent.com/1509205/109728911-1352ea80-7b85-11eb-8d71-856715b27bc4.gif" width="320"></kbd>
  
### Gallery 🔴 

Android | iOS 
--------|-------
<kbd><img src="https://user-images.githubusercontent.com/1509205/109728897-1057fa00-7b85-11eb-9649-f02f6beeef19.gif" width="320"></kbd>    |      <kbd><img src="https://user-images.githubusercontent.com/1509205/109728898-10f09080-7b85-11eb-906e-0a2521cc6232.gif" width="320"></kbd>
 
 🛠 Error Details 🔴
 * To reproduce the error here, add another Gallery block after the successful addition of the first one and you will see this error. 
 <kbd><img src="https://user-images.githubusercontent.com/1509205/112099746-0ae14480-8b72-11eb-9e02-733faf353e08.png" width="320"></kbd>

 * I am thinking that the fact the `clearLastBlockInserted` in the `gallery/edit.js` is causing the issue but I am not sure how to refactor so that the hook calls are consistent. 
----------

### Regression Testing

1. Verify that undo/redo on all media blocks does not trigger the auto-opening of the media options sheet.
2. Verify that selecting the media-related blocks in the canvas does not trigger the auto opening. ( Note : When you select a media-related block on the emulator, the placeholder may receive the touch event and open the media options sheet.)
## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Review

- [ ] mobile review and resolve of Gallery issue
- [ ] Web review and sign off

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jd-alexander committed:
Moved the last block inserted actions from editor to the block-editor

jd-alexander committed:
Moved the last block inserted reducer from editor to the block-editor

jd-alexander committed:
Moved the last block inserted selector from editor to the block-editor

jd-alexander committed:
Fixed es-lint error.

jd-alexander committed:
Moved last block inserted actions test from editor to the block-editor

jd-alexander committed:
Moved last block inserted reducer test from editor to the block-editor

jd-alexander committed:
Moved last block inserted selector test from editor to the block-editor

jd-alexander committed:
Moved all calls to last block inserted from editor to block-editor

jd-alexander committed:
Merge branch 'rnmobile/simplify_image_insertion_flow-redux-changes' into rnmobile/simplify-media-flow-redux-migration

jd-alexander committed:
last block inserter usage in menu native migrated : editor to block-editor

jd-alexander commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -473,8 +473,8 @@ export default compose( [
 		return {
 			wasBlockJustInserted() {
-				const { clearLastBlockInserted } = dispatch( 'core/editor' );
-				const { wasBlockJustInserted } = select( 'core/editor' );
+				const { clearLastBlockInserted } = dispatch( blockEditorStore );
+				const { wasBlockJustInserted } = select( blockEditorStore );
 
 				const result = wasBlockJustInserted( clientId );
 
```
@fluiddot This is the block of code that is causing the issue. What I am trying to accomplish here is, once I get the results of `wasBlockJustInserted` I need to clear the value that marks the current Gallery as the last block inserted so undo/redo and other state changes do not trigger the opening of the sheet again. 

How do you think I can refactor this? I know that the codebase has hooks such as `useSelect` and `useDispatch` and I was wondering if memorization might be able to help here but I wasn't sure. Let me know what you think.

fluiddot replied:
Yeah, I think the best option here would be to apply a refactor. I posted a [comment](https://github.com/WordPress/gutenberg/pull/30123#issuecomment-805024819) related to this.

fluiddot commented:
>Error Details 🔴
>
>To reproduce the error here, add another Gallery block after the successful addition of the first one and you will see this error.
>
>I am thinking that the fact the clearLastBlockInserted in the gallery/edit.js is causing the issue but I am not sure how to refactor so that the hook calls are consistent.

I think the error is not really caused by the changes of this PR but uncovered an issue that was introduced in an older [PR](https://github.com/WordPress/gutenberg/pull/27573). As you can see in this [line](https://github.com/WordPress/gutenberg/pull/27573/files#diff-7be1df22c250a7378a0796b6c68da7fc1178a98d9e88ab9f59868a6d0b8c467fR417), we're using `useMemo` within `withSelect` function but that hook should be only called within a functional component or a custom hook. I noticed this when I was debugging the app, it was showing me this error on every render:
```
Invariant Violation: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
```

You can verify that the error goes away if you remove the `useMemo`.

## How to fix it

Following your [comment](https://github.com/WordPress/gutenberg/pull/30123#discussion_r599289979), the best idea in my opinion would be to refactor the `withSelect` and `withDispatch` code and use hooks instead.

I did a quick modification and I think this solution should work, you can check the code diff below and if you agree I can push these changes to this branch.

<details><summary>Patch diff</summary>

```
diff --git a/packages/block-library/src/gallery/edit.js b/packages/block-library/src/gallery/edit.js
index a57ed5519ffe79ecfac89643586e9bd581a2ea61..d296542026cbe1f33a05f1c1ff4cd73591851ccd 100644
--- a/packages/block-library/src/gallery/edit.js
+++ b/packages/block-library/src/gallery/edit.js
@@ -31,10 +31,10 @@ import {
 	useBlockProps,
 	store as blockEditorStore,
 } from '@wordpress/block-editor';
-import { Platform, useEffect, useState, useMemo } from '@wordpress/element';
+import { Platform, useEffect, useState } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
 import { getBlobByURL, isBlobURL, revokeBlobURL } from '@wordpress/blob';
-import { useDispatch, withSelect, withDispatch } from '@wordpress/data';
+import { useDispatch, useSelect } from '@wordpress/data';
 import { withViewportMatch } from '@wordpress/viewport';
 import { View } from '@wordpress/primitives';
 import { store as coreStore } from '@wordpress/core-data';
@@ -74,14 +74,11 @@ const MOBILE_CONTROL_PROPS_RANGE_CONTROL = Platform.select( {
 function GalleryEdit( props ) {
 	const {
 		attributes,
+		clientId,
 		isSelected,
 		noticeUI,
 		noticeOperations,
-		mediaUpload,
-		imageSizes,
-		resizedImages,
 		onFocus,
-		wasBlockJustInserted,
 	} = props;
 	const {
 		columns = defaultColumnsNumber( attributes ),
@@ -96,6 +93,67 @@ function GalleryEdit( props ) {
 		blockEditorStore
 	);
 
+	const { imageSizes, mediaUpload, wasBlockJustInserted } = useSelect(
+		( select ) => {
+			const settings = select( blockEditorStore ).getSettings();
+
+			return {
+				imageSizes: settings.imageSizes,
+				mediaUpload: settings.mediaUpload,
+				wasBlockJustInserted: select(
+					blockEditorStore
+				).wasBlockJustInserted( clientId ),
+			};
+		}
+	);
+
+	const { resizedImages } = useSelect(
+		( select ) => {
+			const getMedia = select( coreStore ).getMedia;
+
+			if ( isSelected ) {
+				return reduce(
+					attributes.ids,
+					( currentResizedImages, id ) => {
+						if ( ! id ) {
+							return currentResizedImages;
+						}
+						const image = getMedia( id );
+						const sizes = reduce(
+							imageSizes,
+							( currentSizes, size ) => {
+								const defaultUrl = get( image, [
+									'sizes',
+									size.slug,
+									'url',
+								] );
+								const mediaDetailsUrl = get( image, [
+									'media_details',
+									'sizes',
+									size.slug,
+									'source_url',
+								] );
+								return {
+									...currentSizes,
+									[ size.slug ]:
+										defaultUrl || mediaDetailsUrl,
+								};
+							},
+							{}
+						);
+						return {
+							...currentResizedImages,
+							[ parseInt( id, 10 ) ]: sizes,
+						};
+					},
+					{}
+				);
+			}
+			return {};
+		},
+		[ isSelected, attributes.ids, imageSizes ]
+	);
+
 	function setAttributes( newAttrs ) {
 		if ( newAttrs.ids ) {
 			throw new Error(
@@ -345,7 +403,7 @@ function GalleryEdit( props ) {
 			notices={ hasImages ? undefined : noticeUI }
 			onFocus={ onFocus }
 			autoOpenMediaUpload={
-				! hasImages && isSelected && wasBlockJustInserted()
+				! hasImages && isSelected && wasBlockJustInserted
 			}
 		/>
 	);
@@ -417,75 +475,6 @@ function GalleryEdit( props ) {
 }
 
 export default compose( [
-	withSelect( ( select, { attributes: { ids }, isSelected } ) => {
-		const { getMedia } = select( coreStore );
-		const { getSettings } = select( blockEditorStore );
-		const { imageSizes, mediaUpload } = getSettings();
-
-		const resizedImages = useMemo( () => {
-			if ( isSelected ) {
-				return reduce(
-					ids,
-					( currentResizedImages, id ) => {
-						if ( ! id ) {
-							return currentResizedImages;
-						}
-						const image = getMedia( id );
-						const sizes = reduce(
-							imageSizes,
-							( currentSizes, size ) => {
-								const defaultUrl = get( image, [
-									'sizes',
-									size.slug,
-									'url',
-								] );
-								const mediaDetailsUrl = get( image, [
-									'media_details',
-									'sizes',
-									size.slug,
-									'source_url',
-								] );
-								return {
-									...currentSizes,
-									[ size.slug ]:
-										defaultUrl || mediaDetailsUrl,
-								};
-							},
-							{}
-						);
-						return {
-							...currentResizedImages,
-							[ parseInt( id, 10 ) ]: sizes,
-						};
-					},
-					{}
-				);
-			}
-			return {};
-		}, [ isSelected, ids, imageSizes ] );
-
-		return {
-			imageSizes,
-			mediaUpload,
-			resizedImages,
-		};
-	} ),
-	withDispatch( ( dispatch, { clientId }, { select } ) => {
-		return {
-			wasBlockJustInserted() {
-				const { clearLastBlockInserted } = dispatch( blockEditorStore );
-				const { wasBlockJustInserted } = select( blockEditorStore );
-
-				const result = wasBlockJustInserted( clientId );
-
-				if ( result ) {
-					clearLastBlockInserted();
-					return true;
-				}
-				return false;
-			},
-		};
-	} ),
 	withNotices,
 	withViewportMatch( { isNarrow: '< small' } ),
 ] )( GalleryEdit );
```
</details>

These changes shouldn't revert the performance improvement that was done in the [PR](https://github.com/WordPress/gutenberg/pull/27573) because for calculating the `resizedImages` object, I'm passing the same deps array to the `useSelect` function that [should memoize the value](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#useSelect).

jd-alexander commented:
Thanks for the review @fluiddot 🙇🏾 
> I think the error is not really caused by the changes of this PR but uncovered an issue that was introduced in an older PR. As you can see in this line, we're using useMemo within withSelect function but that hook should be only called within a functional component or a custom hook. I noticed this when I was debugging the app, it was showing me this error on every render:

Indeed! I saw that error as well. Happy to see that you narrowed it to down to the `useMemo` usage here. I knew it had something to do with memoization 😄 

> Following your comment, the best idea in my opinion would be to refactor the withSelect and withDispatch code and use hooks instead.

Sounds good! 

> I did a quick modification and I think this solution should work, you can check the code diff below and if you agree I can push these changes to this branch.

Thanks! Yes, I reviewed the diff and I can see clearly where the code is migrated from the high order components to hooks. 

The only logic missing from the diff is that of `clearLastBlockInserted` that needs to be dispatched (`useDispatch`) so that the state for the last block inserted is cleared once the bottom sheet is shown. Without this, when the user performs an undo/redo operation et. al the bottom sheet will reopen unexpectedly for the block. 

> These changes shouldn't revert the performance improvement that was done in the PR because for calculating the resizedImages object, I'm passing the same deps array to the useSelect function that should memoize the value.

Makes sense. I think this migration is pretty straightforward 👍🏾 A sanity test of the Gallery block will be done here to verify all is well once the changes are settled.

fluiddot commented:
>The only logic missing from the diff is that of clearLastBlockInserted that needs to be dispatched (useDispatch) so that the state for the last block inserted is cleared once the bottom sheet is shown. Without this, when the user performs an undo/redo operation et. al the bottom sheet will reopen unexpectedly for the block.

Oh I didn't realise about the undo/redo operations, actually I removed the `clearLastBlockInserted` call because I thought it wasn't necessary.
 
Regarding this part, I'm thinking if there would be a better approach. I wouldn't expect that an edit component clears the value of the last block inserted when the bottom sheet is displayed. We might want that value in the future for a different logic that is executed later on, but in this case, it might be cleared before fetching it.
 
I'm wondering if it would be possible to pass a prop with the insertion source (`inserter_menu`, `clipboard`, etc.) to the edit component, this way we would know if it was inserted by the inserter menu and therefore we could display the picker. What do you think?

jd-alexander commented:
> Regarding this part, I'm thinking if there would be a better approach. I wouldn't expect that an edit component clears the value of the last block inserted when the bottom sheet is displayed. We might want that value in the future for a different logic that is executed later on, but in this case, it might be cleared before fetching it.
I'm wondering if it would be possible to pass a prop with the insertion source (inserter_menu, clipboard, etc.) to the edit component, this way we would know if it was inserted by the inserter menu and therefore we could display the picker. What do you think?

Sounds like a good idea. I am wondering though if this solution would prevent re-renders of the auto opening. i.e if the insertion source prop gets passed to the component and we check if a block was last inserted, what happens when there are re-renders? how would we track that the bottom sheet is already shown? Currently clearing it does that but I agree with the fact that if the last block inserted functionality was needed elsewhere clearing it would break that functionality.

fluiddot commented:
>I am wondering though if this solution would prevent re-renders of the auto opening. i.e if the insertion source prop gets passed to the component and we check if a block was last inserted, what happens when there are re-renders? 

As far as I checked re-renders are not a problem unless `wasBlockJustInserted` changes from `false` to `true`, if so the picker is displayed.

>how would we track that the bottom sheet is already shown?

I think we don't really need to track this, in my opinion the part that we have to control is the `wasBlockJustInserted`.

> Currently clearing it does that but I agree with the fact that if the last block inserted functionality was needed elsewhere clearing it would break that functionality.

I think clearing the last block inserted makes sense but we should trigger it in a different moment.

I explored the option of passing a source when a block is inserted and I think it works fine, you can check the changes in this [PR](https://github.com/WordPress/gutenberg/pull/30238). Let me know your feedback, if you think the changes are ok, we could apply it to this PR.

fluiddot committed:
[RNMobile] Refactor: Simplify media flow redux migration (#30238)

* Add meta argument to insertBlock action

* Add inserter menu source

* Update last block inserted reducer

Instead of using specific actions for tracking the last block inserted, it uses the actions related to the insertion flow.

* Add get last block inserted selector

* Refactor gallery edit component

withSelect and withDispatch logic has been refactored to use useSelect and useDispatch hooks

* Refactor image edit component

wasBlockJustInserted is now calculated with getLastBlockInserted

* Refactor video edit component

wasBlockJustInserted is now calculated with getLastBlockInserted

* Fix reset blocks action in last block inserted reducer

* Add source param to wasBlockJustInserted selector

* Simplify withSelect part of video block

jd-alexander committed:
Removed add/clear last block inserted actions and tests due to refactor

jd-alexander committed:
Removed addLastBlockInserted from the insert menu's onPress.

jd-alexander committed:
rewrote the tests for the lastBlockInserted reducer.

jd-alexander committed:
rewrote tests for wasBlockJustInserted selector.

jd-alexander committed:
optimized clientId

jd-alexander committed:
removed unneeded clientId.

jd-alexander committed:
put the expectedSource inside the test meta object.

jd-alexander committed:
used expectedState instead {} for state related tests.

jd-alexander committed:
used expectedState insted {} for state related tests.

jd-alexander committed:
Merge remote-tracking branch 'origin/rnmobile/simplify-media-flow-redux-migration' into rnmobile/simplify-media-flow-redux-migration

jd-alexander commented:
Hi @fluiddot this is ready for review. 

The changes included are as follows : 
1. Removed the `CLEAR_LAST_BLOCK_INSERTED` and `ADD_LAST_BLOCK_INSERTED` actions, reducers, selectors.
2. Removed the `CLEAR_LAST_BLOCK_INSERTED` and `ADD_LAST_BLOCK_INSERTED` tests.
3. Created unit tests for the different paths related to the `INSERT_BLOCKS` and `RESET_BLOCKS` actions in the reducer.
4. The `lastBlockInserted` selector tests now utilize the new insertion source approach - `inserter_menu`.

fluiddot commented on packages/block-editor/src/store/test/reducer.js:
```diff
@@ -2944,4 +2945,98 @@ describe( 'state', () => {
 		} );
 	} );
+
+	describe( 'lastBlockInserted()', () => {
```
I think we can remove the parenthesis (`()`) from the describe name.

jd-alexander replied:
Noted. Removed here ae46525 I also removed the parenthesis (`()`) from another test that I used as an example that contained its usage.

fluiddot commented on packages/block-editor/src/store/test/reducer.js:
```diff
@@ -2944,4 +2945,98 @@ describe( 'state', () => {
+		} );
+
+		it( 'should return empty state if last block inserted is called with action INSERT_BLOCKS that is not a updateSelection', () => {
+			const expectedState = {};
```
This test is correct but I think it's not 100% accurate, if the `INSERT_BLOCKS` action has the param `updateSelection` as `false`, the reducer returns the same state, not an empty state.

We could validate this if `expectedState` is not an empty object:
```
const expectedState = {
	clientId: '9db792c6-a25a-495d-adbd-97d56a4c4189',
};
```

jd-alexander replied:
Good catch! Resolved here 5c1d9fc

fluiddot commented on packages/block-editor/src/store/test/reducer.js:
```diff
@@ -2944,4 +2945,98 @@ describe( 'state', () => {
+		} );
+
+		it( 'should return empty state if last block inserted is called with action INSERT_BLOCKS and actions.blocks = 0', () => {
+			const expectedState = {};
```
Here we have the same situation as in the previous comment:
> This test is correct but I think it's not 100% accurate, if the INSERT_BLOCKS action has the param updateSelection as false, the reducer returns the same state, not an empty state.

fluiddot replied:
I'd also update the last part of the test description, instead of `actions.blocks = 0`, we could say `blocks list is empty`, wdyt?

jd-alexander replied:
No problem. I resolved this in the same commit as above 5c1d9fc and I updated the test description as well 
9307bf6

fluiddot commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -2084,3 +2084,19 @@ export const __experimentalGetActiveBlockIdByBlockNames = createSelector(
+ *
+ * @param {Object} state Global application state.
+ * @param {Object} clientId Client Id of the block.
+ * @param {boolean} source Insertion source of the block.
```
The type of this param should be `string` (sorry because I missed this when I made the changes 😄 ).

fluiddot replied:
Regarding my previous comment, what do you think about setting the `source` param as optional?
In most of cases the blocks are inserted without a source so if we want to know if a block was inserted from an unknown source, we could do `wasBlockJustInserted(SOME_CLIENT_ID)` instead of `wasBlockJustInserted(SOME_CLIENT_ID, undefined)`.

jd-alexander replied:
Agreed. I marked the source as optional and updated the type of the param 0d0164c Just to be clear since function parameters default to `undefined` it's fine to just update the documentation and leave the parameter as is?

fluiddot replied:
>Just to be clear since function parameters default to undefined it's fine to just update the documentation and leave the parameter as is?

Yeah I think so, if the default value would be different than `undefined` (here is an [example](https://github.com/WordPress/gutenberg/blob/d2e3b2b6cf8e681938403817422bbfa57c95a97d/packages/notices/src/store/selectors.js#L45-L56)) then we should set it in the param. In this case since the default is `undefined`, I think it's ok to leave it this way.

jd-alexander committed:
removed parentheses from describe name.

jd-alexander committed:
return the same state instead of empty state.

jd-alexander committed:
made changes to test name so its intent is clearer.

jd-alexander committed:
made the insertion source optional.

fluiddot reviewed (approved):
LGTM 🎊 !

_Tested on iPhone 11 and Samsung Galaxy S20 FE 5G._
